content/frontend-web/react-typescript/design-pattern.md

```diff
@@ -6,6 +6,217 @@ showToc: true
+
+- Default export
+
+    ```ts
+    // MyComponent.tsx
+    const MyComponent = () => <div>Hey, MyComponent</div>;
+    export default MyComponent;
+    // App.tsx
+    import MyComponent from './MyComponent';
+    ```
+
+- Export
+
+    ```ts
+    // MyComponent.tsx
+    export const MyComponent = () => <div>Hey, MyComponent</div>;
+    // App.tsx
+    import { MyComponent } from './MyComponent';
+    ```
+
+- Rename import
+
+    ```ts
+    // MyComponent.tsx
+    export const MyComponent = () => <div>Hey, MyComponent</div>;
+    // App.tsx
+    import { MyComponent as MySuperComponent } from './MyComponent';
+    export const App = () => <MySuperComponent />;
+    ```
+
+- Group all imports and rename
+
+    ```tsx
+    // MyComponent.tsx
+    export const Switch = () => <div>Hey, Switch</div>;
+    export const Group = () => <div>Hey, Group</div>;
+    // App.tsx
+    import { * as MyComponent } from './MyComponent';
+    export const App = () => (
+        <>
+            <MyComponent.Switch />
+            <MyComponent.Group />
+        </>
+    );
+    ```
+
+- Re-export all
+
+    ```ts
+    // MyComponent.tsx
+    export const Switch = () => <div>Hey, Switch</div>;
+    export const Group = () => <div>Hey, Group</div>;
+    // MyParentComponent.tsx
+    export * from "./MyComponent";
+    // App.tsx
+    import { Switch, Group } from './MyComponent';
+    export const App = () => (
+        <>
+            <Switch />
+            <Group />
+        </>
+    );
+    ```
+
+- Re-export, then group all imports and rename
+
+    ```ts
+    // MyComponent.tsx
+    export const Switch = () => <div>Hey, Switch</div>;
+    export const Group = () => <div>Hey, Group</div>;
+    // MyParentComponent.tsx
+    export  * as MyComponent from "./MyComponent";
+    // App.tsx
+    import { MyComponent } from './MyComponent';
+    export const App = () => (
+        <>
+            <MyComponent.Switch />
+            <MyComponent.Group />
+        </>
+    );
+    ```
+
+
+- Export type
+
+    ```ts
+    // MyComponent.tsx
+    export type Sex = "Male" | "Female"; // Accept only "Male" or "Female" 
+    // App.tsx
+    import { Sex } from './MyComponent'
+    ```
+
+- Export interface
+
+    ```ts
+    // MyComponent.tsx
+    export interface User {
+        id: string;
+        name: string;
+    }
+    // App.tsx
+    import { User } from './MyComponent'
+    ```
+
+
+# Handling with side effects
+
+Side effects เป็น function โดยส่วนใหญ่แล้ว เราจะจัดการกับ Side Effect ของ Component โดยมี 3 scenarios:
+- ทำงานเฉพาะครั้งแรกที่ Component mount (Only when a component mounts)
+- เคลียร์ค่าต่างๆ โดยใช้ return a function (Cleaning up by returning a function)
+- ทำงานเมื่อ dependencies มีการเปลี่ยนแปลง (Running when specific dependencies change)a
+
+## ทำงานเฉพาะครั้งแรกที่ Component mount
+
+จะทำครั้งแรก ครั้งเดียวเมื่อ Component โหลดเสร็จ
+
+```ts
+function FirstExecuteComponent() {
+
+  // add event listeners (Flux Store, WebSocket, document, etc.)

```

Comment:
Very minor, please prefer to start a comment with A capital alphabet like a normal sentence.